### PR TITLE
Add NEO Support

### PR DESCRIPTION
**Pull request by newcodepusher**
## Description

Adding support for NEO blockchain
https://neo.org/

## Testing instructions

This PR includes unit tests and swift/kotlin integration tests

## Types of changes
* New feature (non-breaking change which adds functionality)

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

**Comment by hewigovens** [comment]
Hi, could you please create a clean pr from master? too many commits in history

**Comment by hewigovens** [comment]
can you also remove `TWNEOAddress.h/cpp` and adopt `TWAnyAddress`?

**Comment by newcodepusher** [comment]
Hi, all changes have been done

File: tests/NEO/AccountTests.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#include <gtest/gtest.h>
+
+using namespace TW;
+using namespace TW::NEO;
+
+TEST(NEOAccount, validity) {
+    auto hexPrvKey = "4646464646464646464646464646464646464646464646464646464646464646";
+    auto hexPubKey = "031bec1250aa8f78275f99a6663688f31085848d0ed92f1203e447125f927b7486";
+    auto signer = Signer(PrivateKey(parse_hex(hexPrvKey)));
+    auto prvKey = signer.getPrivateKey();
```
Comment (hewigovens) [inline tests/NEO/AccountTests.cpp]: let's remove these methods and test from Signer

File: src/NEO/CoinReference.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    virtual ~CoinReference() {}
+
+    int64_t size() const override {
+        return 32 + 2; // sizeof(prevHash) + sizeof(prevIndex);
+    }
+
+    void deserialize(const Data &data, int initial_pos = 0) override {
+        prevHash = load<uint256_t>(readBytes(data, 32, initial_pos));
+        prevIndex = decode16LE(data.data() + initial_pos + 32);
```
Comment (catenocrypt) [inline src/NEO/CoinReference.h]: 32: please put this const magic number into a named const

File: src/NEO/Script.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#include "OpCode.h"
+
+namespace TW::NEO {
+
+Data Script::CreateSignatureRedeemScript(Data publicKey) {
+  Data result;
+  result.push_back((byte)PUSHBYTES21);
+  result.insert(result.end(), publicKey.begin(), publicKey.end());
+  result.push_back((byte)CHECKSIG);
+  return result;
```
Comment (catenocrypt) [inline src/NEO/Script.cpp]: Why only 2 space indent here?  In other place 4 is used.

File: src/NEO/Signer.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#include "Address.h"
+#include "Transaction.h"
+
+namespace TW::NEO {
+
+class Signer {
+  private:
+    Data publicKey;
+    TW::PrivateKey privateKey;
+    Address* address;
```
Comment (catenocrypt) [inline src/NEO/Signer.h]: Why is a pointer needed here?  Why couldn't it be value type?

File: src/NEO/ISerializable.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+#pragma once
+
+#include "../Data.h"
+#include "../BinaryCoding.h"
+#include "ReadData.h"
+
+namespace TW::NEO {
+
+class ISerializable {
```
Comment (catenocrypt) [inline src/NEO/ISerializable.h]: Please move the static methods into a separate class, to keep the interface a real interface.

File: src/NEO/Address.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    std::copy(keyHash.data(), keyHash.data() + Address::size, bytes.begin());
+}
+
+Address::Address(uint8_t m, const std::vector<Data>& publicKeys) {
+    auto builderData = toScriptHash(Ontology::ParamsBuilder::fromMultiPubkey(m, publicKeys));
+    std::copy(builderData.begin(), builderData.end(), bytes.begin());
+}
+
+Data Address::toScriptHash(const Data& data) const {
+    return Hash::ripemd(Hash::sha256(data));
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply src/NEO/Address.cpp]: yes, i think this is a useful method

File: coins.json [hunk 7]
```diff
@@ -1306,5 +1306,28 @@
...
             "clientDocs": "https://cardanodocs.com/introduction/"
         }
+    },
+    {
+        "id": "neo",
+        "name": "NEO",
+        "symbol": "NEO",
+        "decimals": 8,
+        "blockchain": "NEO",
+        "derivationPath": "m/44'/888'/0'/0'/0'",
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply coins.json]: yes, you right, i fixed it

File: coins.json [hunk 8]
```diff
@@ -1306,5 +1306,28 @@
...
+            "url": "https://neoscan.io",
+            "txPath": "/transaction/",
+            "accountPath": "/address/",
+            "sampleTx": "e0ddf7c81c732df26180aca0c36d5868ad009fdbbe6e7a56ebafc14bba41cd53",
+            "sampleAccount": "AcxuqWhTureEQGeJgbmtSWNAtssjMLU7pb"
+        },
+        "info": {
+            "url": "https://neo.org",
+            "client": "https://github.com/neo-project/neo",
+            "clientPublic": "",
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply coins.json]: added, but not official RPC, just stable

File: include/TrustWalletCore/TWNEOProto.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
+// Copyright © 2017-2020 Trust Wallet.
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply include/TrustWalletCore/TWNEOProto.h]: yes, removed it

File: src/NEO/Address.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+bool Address::isValid(const std::string& string) {
+    const auto decoded = Base58::bitcoin.decodeCheck(string);
+    return !(decoded.size() != Address::size || decoded[0] != version);
+}
+
+Address::Address(const PublicKey& publicKey) {
+    auto publicKeyData = publicKey.bytes;
+
+    auto pkdata = Data(publicKeyData.begin(), publicKeyData.end());
+    pkdata.insert(pkdata.begin(), 0x21);
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply src/NEO/Address.cpp]: done

File: src/NEO/Signer.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  return privateKey;
+}
+
+TW::PublicKey Signer::getPublicKey() const {
+  return TW::PublicKey(publicKey, TWPublicKeyTypeNIST256p1);
+}
+
+Address Signer::getAddress() const {
+  return Address(address);
+}
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply src/NEO/Signer.cpp]: yes, i think this is a useful method

File: src/NEO/Signer.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    explicit Signer(const TW::PrivateKey &priKey);
+    PrivateKey getPrivateKey() const;
+    PublicKey getPublicKey() const;
+    Address getAddress() const;
+
+    void sign(Transaction &tx) const;
+    Data sign(const Data &data) const;
+
+    /// Signs a Proto::SigningInput transaction
+    //static Proto::SigningOutput sign(const Proto::SigningInput& input) noexcept;
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply src/NEO/Signer.h]: done

File: src/NEO/Signer.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+#include "Address.h"
+#include "Transaction.h"
+
+namespace TW::NEO {
+
+class Signer {
+  private:
+    Data publicKey;
+    TW::PrivateKey privateKey;
+    std::string address;
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply src/NEO/Signer.h]: done

File: tests/NEO/AccountTests.cpp (hunk 1 above)
Comment (newcodepusher) [reply tests/NEO/AccountTests.cpp]: done

File: src/NEO/Signer.h (hunk 4 above)
Comment (newcodepusher) [reply src/NEO/Signer.h]: I didn’t want to add a default constructor with an initialization like Address(«1111111111111111111114oLvT2»). I start extend address class for this and remove pointer

File: src/NEO/ISerializable.h (hunk 5 above)
Comment (newcodepusher) [reply src/NEO/ISerializable.h]: done

File: src/NEO/Script.cpp (hunk 3 above)
Comment (newcodepusher) [reply src/NEO/Script.cpp]: fixed

File: src/NEO/CoinReference.h (hunk 2 above)
Comment (newcodepusher) [reply src/NEO/CoinReference.h]: doen

**Comment by newcodepusher** [comment]
Hi, I was testing linux unit tests and all passed.
Something happened with test on server
### docker: error pulling image configuration:

File: src/NEO/Address.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+Data Address::toScriptHash(const Data& data) const {
+    return Hash::ripemd(Hash::sha256(data));
+}
+
+Data Address::toScriptHash() const {
+	 byte buf[20];
+	 Data data(buf, buf + 20);
+	 std::copy(bytes.begin() + 1, bytes.begin() + 21, data.begin());
+    return data;
```
[the remark this answers is not in the document]
Comment (newcodepusher) [reply src/NEO/Address.cpp]: done

File: include/TrustWalletCore/TWNEOSigner.h [hunk 15]
```diff
@@ -0,0 +1,21 @@
+// Copyright © 2017-2019 Trust Wallet.
```
[the remark this answers is not in the document]
Comment (hewigovens) [reply include/TrustWalletCore/TWNEOSigner.h]: can you search and replace all copyright year? new added files should be 2020

File: src/NEO/Signer.cpp [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+#include "../Hash.h"
+#include "../PublicKey.h"
+#include "../PrivateKey.h"
+#include "Script.h"
+#include "Signer.h"
+
+using namespace TW;
+using namespace TW::NEO;
+
+Signer::Signer(const TW::PrivateKey &priKey)
```
Comment (hewigovens) [inline src/NEO/Signer.cpp]:
```suggestion
Signer::Signer(const PrivateKey &priKey)
```

File: src/NEO/Signer.h [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+#pragma once
+
+#include "../PrivateKey.h"
+#include "../Data.h"
+#include "Address.h"
+#include "Transaction.h"
+
+namespace TW::NEO {
+
+class Signer {
```
Comment (hewigovens) [inline src/NEO/Signer.h]: we're unifying signer interface, can you follow this template? https://github.com/trustwallet/wallet-core/blob/master/codegen/lib/templates/newcoin/Signer.h.erb? the idea is to accept `Proto::SigningInput` so we can implement generic signer later
Thanks

File: src/UInt.h [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace TW {
+
+namespace bm = boost::multiprecision;
+
+using int160_t = bm::number<bm::backends::cpp_int_backend<160, 160, bm::signed_magnitude, bm::unchecked, void>>;
+using uint160_t = bm::number<bm::backends::cpp_int_backend<160, 160, bm::unsigned_magnitude, bm::unchecked, void>>;
+
+template<class T>
+inline T load(const Data &data, int initial_position = 0) {
```
Comment (hewigovens) [inline src/UInt.h]: can we merge this with `uint256.h` together somehow?

File: src/interface/TWNEOSigner.cpp [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+#include <TrustWalletCore/TWNEOSigner.h>
+
+#include "../NEO/Signer.h"
+#include "../proto/NEO.pb.h"
+#include "../HexCoding.h"
+
+using namespace TW;
+using namespace TW::NEO;
+
+TW_NEO_Proto_SigningOutput TWNEOSignerSign(TW_NEO_Proto_SigningInput data) {
```
Comment (hewigovens) [inline src/interface/TWNEOSigner.cpp]: see my last comment about signer, let's move this logic into `Signer` to keep TWNEOSigner as simple as possible, https://github.com/trustwallet/wallet-core/blob/master/codegen/lib/templates/newcoin/TWSigner.cpp.erb

File: src/NEO/Signer.cpp (hunk 16 above)
Comment (newcodepusher) [reply src/NEO/Signer.cpp]: done

File: src/UInt.h (hunk 18 above)
Comment (newcodepusher) [reply src/UInt.h]: done

File: src/interface/TWNEOSigner.cpp (hunk 19 above)
Comment (newcodepusher) [reply src/interface/TWNEOSigner.cpp]: done

File: src/proto/NEO.proto [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    string asset_id = 1;
+    sint64 value = 2;
+    string address = 3;
+}
+
+// Input data necessary to create a signed transaction.
+message SigningInput {
+    repeated TransactionInput inputs = 1;
+    repeated TransactionOutput outputs = 2;
+    bytes private_key = 3;
```
Comment (hewigovens) [inline src/proto/NEO.proto]: Sorry for overlooked about fee and changes: 
can you take a look at Bitcoin.proto or Cardano.proto? how do we calculate fee? for other UTXO blockchains, we will do two phases sign:
1. calculate the fee and return a TransactionPlan
2. the signer accept the plan and generate output automatically

Comment (newcodepusher) [reply src/proto/NEO.proto]: I added the TransactionPlan implementation as in Cardano, but with some corrections for NEO. The fact is, you can transfer NEO or several different tokens at once, but the fee must be paid in GAS.

> Transactions larger than 1024 bytes require 0.001 GAS as a base fee, in addition to 0.00001 GAS per byte of the transaction. Most basic NEP-5 transfers remain well under the 1024 byte threshold, however transactions with more complex logic or large quantities of inputs and outputs may exceed the threshold and require a network fee.

File: src/NEO/Signer.cpp [hunk 21]
```diff
@@ -49,37 +50,157 @@ Data Signer::sign(const Data &data) const {
...
+    for (int i = 0; i < input.inputs_size(); i++) {
+        if (available.find(input.inputs(i).asset_id()) == available.end()) {
+            available[input.inputs(i).asset_id()] = input.inputs(i).value();
+            plan.add_inputs()->MergeFrom(input.inputs(i));
+            continue;
+        }
+
+        if (input.inputs(i).asset_id() != input.gas_asset_id()
+                && required[input.inputs(i).asset_id()] < available[input.inputs(i).asset_id()])
+            continue;
```
Comment (hewigovens) [inline src/NEO/Signer.cpp]: we prefer `if {}`, always with {}

File: src/NEO/Signer.cpp [hunk 22]
```diff
@@ -49,37 +50,157 @@ Data Signer::sign(const Data &data) const {
...
+        plan.add_inputs()->MergeFrom(input.inputs(i));
+    }
+
+    int existGASTransfer = -1;
+    for (int i = 0; i < input.outputs_size(); i++) {
+        auto outputPlan = plan.add_outputs();
+
+        if (available.find(input.inputs(i).asset_id()) == available.end()
+            || available[input.outputs(i).asset_id()] < input.outputs(i).amount())
+            throw "Input balance for asset too low";
```
Comment (hewigovens) [inline src/NEO/Signer.cpp]: can you add `{}` for if statement?

Comment (newcodepusher) [reply src/NEO/Signer.cpp]: done

File: src/NEO/Signer.cpp (hunk 21 above)
Comment (newcodepusher) [reply src/NEO/Signer.cpp]: done

**Pull request merged**